grpc/stream.go

```diff
@@ -0,0 +1,325 @@
+
+	must(rt, s.obj.DefineDataProperty(
+		"end", rt.ToValue(s.end), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+}
+
+func (s *stream) beginStream() error {
+	req := &grpcext.StreamRequest{
+		Method:           s.method,
+		MethodDescriptor: s.methodDescriptor,
+	}
+
+	stream, err := s.client.conn.NewStream(s.vu.Context(), *req, metadata.MD{})
+	if err != nil {
+		return fmt.Errorf("failed to create a new stream: %w", err)
+	}
+	s.stream = stream
+
+	go s.loop()
+
+	return nil
+}
+
+func (s *stream) loop() {
+	ctx := s.vu.Context()
+	wg := new(sync.WaitGroup)
+
+	defer func() {
+		wg.Wait()
+		s.tq.Close()
+	}()
+
+	// read & write data from/to the stream
+	wg.Add(2)
+	go s.readData(wg)
+	go s.writeData(wg)
+
+	ctxDone := ctx.Done()
+	for {
+		select {
+		case <-ctxDone:
+			// VU is shutting down during an interrupt
+			// stream events will not be forwarded to the VU
+			s.tq.Queue(func() error {
+				return s.closeWithError(nil)
+			})
+			return
+		case <-s.done:
+			return
+		}
+	}
+}
+
+func (s *stream) queueMessage(msg map[string]interface{}) {
+	s.tq.Queue(func() error {
+		rt := s.vu.Runtime()
+		listeners := s.eventListeners.all(eventData)
+
+		for _, messageListener := range listeners {
+			if _, err := messageListener(rt.ToValue(msg)); err != nil {
+				// TODO(olegbespalov) consider logging the error
+				_ = s.closeWithError(err)
+
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+// readData reads data from the stream and forward them to the readDataChan
+func (s *stream) readData(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for {
+		msg, err := s.stream.ReceiveConverted()
+
+		if err != nil && !isRegularClosing(err) {
+			s.logger.WithError(err).Debug("error while reading from the stream")
+
+			s.tq.Queue(func() error {
+				return s.closeWithError(err)
+			})
+
+			return
+		}
+
+		if len(msg) == 0 && isRegularClosing(err) {
+			s.logger.WithError(err).Debug("stream is cancelled/finished")
+
+			s.tq.Queue(func() error {
+				return s.closeWithError(nil)
+			})
+
+			return
+		}
+
+		s.queueMessage(msg)
+	}
+}
+
+func isRegularClosing(err error) bool {
+	return errors.Is(err, io.EOF) || errors.Is(err, grpcext.ErrCanceled)
+}
+
+// writeData writes data to the stream
+func (s *stream) writeData(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	writeChannel := make(chan message)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case msg, ok := <-writeChannel:
+				if !ok {
+					return
+				}
+
+				err := s.stream.Send(msg.msg)
+				if err != nil {
+					s.logger.WithError(err).Error("failed to send data to the stream")
+
+					s.tq.Queue(func() error {
+						closeErr := s.closeWithError(err)
+						return closeErr
+					})
+					return
+				}
+			case _, ok := <-s.closeSend:
+				if !ok {
+					continue
+				}
+
+				if err := s.stream.CloseSend(); err != nil {
+					s.logger.WithError(err).Error("an error happened during stream closing")
+				}
+
+				close(s.closeSend)
```

Comment:
I somehow missed this, but ... uh.

In general writers close channels, not readers. THis also will loop and will keep going in the `!ok` branch and will go in the select and enter again ... and again.

THis is why I do set the channel to `nil` in the websocket code base.

Reply:
:thinking: right, moreover end could be not called. So let me rethink the approach

Reply:
I don't thknk `end` not being called is a valid code ata all. There is no way for us to know whether a `stream` should end if it is not called or should continue longer.

So it is expected that if `end` is not called that k6 will just wait for it indefinitely ... except I guess if there is an error that closes the stream :shrug:

Reply:
>I don't thknk end not being called is a valid code ata all. There is no way for us to know whether a stream should end if it is not called or should continue longer.

It's for the server-side streaming only case 

>except I guess if there is an error that closes the stream 

Exactly, in that case, we should rely only on fore the `io.EOF` during the reading of the stream